src/ui/activity-iframe-view.ts

```diff
@@ -0,0 +1,169 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ActivityIframePort, ActivityPorts} from '../components/activities';
+import {ActivityResult} from 'web-activities/activity-ports';
+import {Dialog} from '../components/dialog';
+import {Message} from '../proto/api_messages';
+import {View} from '../components/view';
+import {acceptPortResultData} from '../utils/activity-utils';
+import {createElement} from '../utils/dom';
+import {isCancelError} from '../utils/errors';
+
+const iframeAttributes = {
+  'frameborder': '0',
+  'scrolling': 'no',
+};
+
+/**
+ * Class to build and render Activity iframe view.
+ */
+export class ActivityIframeView extends View {
+  private readonly doc_: Document;
+  private readonly iframe_: HTMLIFrameElement;
+  private port_: ActivityIframePort | null = null;
+  private portResolver_?: (
+    port: ActivityIframePort | Promise<ActivityIframePort>
+  ) => void;
+  private readonly portPromise_: Promise<ActivityIframePort>;
+
+  constructor(
+    private readonly win_: Window,
+    private readonly activityPorts_: ActivityPorts,
+    private readonly src_: string,
+    /** Additional data to be passed to the iframe. */
+    private readonly args_: {[key: string]: string},
+    private readonly shouldFadeBody_: boolean = false,
+    private readonly hasLoadingIndicator_: boolean = false
+  ) {
+    super();
+
+    this.doc_ = this.win_.document;
+
+    this.iframe_ = createElement(this.doc_, 'iframe', iframeAttributes);
+
+    this.portPromise_ = new Promise((resolve) => {
+      this.portResolver_ = resolve;
+    });
+  }
+
+  getElement() {
+    return this.iframe_;
+  }
+
+  async init(dialog: Dialog) {
+    const port = await this.activityPorts_.openIframe(
+      this.iframe_,
+      this.src_,
+      this.args_
+    );
+    return this.onOpenIframeResponse_(port, dialog);
+  }
+
+  /**
+   * Returns if document should fade for this view.
+   */
+  shouldFadeBody(): boolean {
+    return this.shouldFadeBody_;
+  }
+
+  /**
+   * Returns if the view shows loading indicator.
+   */
+  hasLoadingIndicator(): boolean {
+    return this.hasLoadingIndicator_;
+  }
+
+  private onOpenIframeResponse_(
+    port: ActivityIframePort,
+    dialog: Dialog
+  ): Promise<void> {
+    this.port_ = port;
+    this.portResolver_!(port);
+
+    this.port_.onResizeRequest((height) => {
+      dialog.resizeView(this, height);
+    });
+
+    return this.port_.whenReady();
+  }
+
+  private getPortPromise_(): Promise<ActivityIframePort> {
+    return this.portPromise_;
+  }
+
+  async on<T extends Message>(
+    message: new (data?: unknown[], includesLabel?: boolean) => T,
```

Comment:
Should this use the `MessageConstructor` type?

Reply:
I think this template approach could work better.

We don't export the `MessageConstructor` type currently. I tried recreating it and calling it from an example method, and ran into an issue (screenshots below).
![image](https://user-images.githubusercontent.com/1216762/228084401-49eee1cf-a401-49c1-9058-a54a3307338b.png)
![image](https://user-images.githubusercontent.com/1216762/228084615-0e26a136-2191-4dd1-b245-280d0663e922.png)


Whereas if I stick with the template approach, the example method works (screenshot below).
![image](https://user-images.githubusercontent.com/1216762/228084511-aab26844-2e5e-4d29-9217-e68db57d2096.png)